File: drills/ts/harnesses/nc146_validate_bst.ts
import * as solution from "./solution";

const show = (v: unknown) => JSON.stringify(v) ?? "undefined";

// LeetCode's level-order form: an array of values with null for a missing
// child, trailing nulls trimmed.
const build = (values: (number | null)[]): any => {
  if (values.length === 0 || values[0] === null) return null;
  const root = new solution.TreeNode(values[0]!);
  const queue: any[] = [root];
  let head = 0;
  let i = 1;
  while (head < queue.length && i < values.length) {
    const node = queue[head++];
    if (i < values.length) {
      const value = values[i++];
      if (value !== null) {
        node.left = new solution.TreeNode(value);
        queue.push(node.left);
      }
    }
    if (i < values.length) {
      const value = values[i++];
      if (value !== null) {
        node.right = new solution.TreeNode(value);
        queue.push(node.right);
      }
    }
  }
  return root;
};

const levels = (node: any): (number | null)[] => {
  const out: (number | null)[] = [];
  const queue: any[] = [node];
  let head = 0;
  while (head < queue.length && out.length < 500) {
    const current = queue[head++];
    if (current === null || current === undefined) {
      out.push(null);
    } else {
      out.push(current.val);
      queue.push(current.left ?? null);
      queue.push(current.right ?? null);
    }
  }
  while (out.length > 0 && out[out.length - 1] === null) out.pop();
  return out;
};

export function run(): [string, string, string][] {
  if (typeof solution.isValidBST !== "function" || typeof solution.TreeNode !== "function") throw new Error("__signature_mismatch__");
  return [
    ["isValidBST([2,1,3])", show(true), show(solution.isValidBST(build([2, 1, 3])))],
    ["isValidBST([])", show(true), show(solution.isValidBST(build([])))],
    ["isValidBST([1])", show(true), show(solution.isValidBST(build([1])))],
    ["isValidBST([5,1,4,null,null,3,6])", show(false), show(solution.isValidBST(build([5, 1, 4, null, null, 3, 6])))],
    ["isValidBST(every node beats its parent, but 3 is on the wrong side)", show(false), show(solution.isValidBST(build([5, 4, 6, null, null, 3, 7])))],
    ["isValidBST([2,2]) -- equal values are not allowed", show(false), show(solution.isValidBST(build([2, 2])))],
  ];
}

File: drills/python/harnesses/nc146_validate_bst.py
try:
    (isValidBST, TreeNode)
except NameError:
    raise Exception("__signature_mismatch__")

__results__ = []
def __case__(label, expected, actual):
    __results__.append([label, repr(expected), repr(actual)])

# LeetCode's level-order form: a list of values with None for a missing child,
# trailing Nones trimmed.
def __build__(values):
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue, head, i = [root], 0, 1
    while head < len(queue) and i < len(values):
        node = queue[head]
        head += 1
        if i < len(values):
            value = values[i]
            i += 1
            if value is not None:
                node.left = TreeNode(value)
                queue.append(node.left)
        if i < len(values):
            value = values[i]
            i += 1
            if value is not None:
                node.right = TreeNode(value)
                queue.append(node.right)
    return root


def __levels__(node):
    out, queue, head = [], [node], 0
    while head < len(queue) and len(out) < 500:
        current = queue[head]
        head += 1
        if current is None:
            out.append(None)
        else:
            out.append(current.val)
            queue.append(current.left)
            queue.append(current.right)
    while out and out[-1] is None:
        out.pop()
    return out

__case__("isValidBST([2,1,3])", True, isValidBST(__build__([2, 1, 3])))
__case__("isValidBST([])", True, isValidBST(__build__([])))
__case__("isValidBST([1])", True, isValidBST(__build__([1])))
__case__("isValidBST([5,1,4,None,None,3,6])", False, isValidBST(__build__([5, 1, 4, None, None, 3, 6])))
__case__("isValidBST(every node beats its parent, but 3 is on the wrong side)", False, isValidBST(__build__([5, 4, 6, None, None, 3, 7])))
__case__("isValidBST([2,2]) -- equal values are not allowed", False, isValidBST(__build__([2, 2])))
